experimental/Origami/src/generate.jl

```diff
@@ -0,0 +1,412 @@
+end
+
+# lengths are realizable if they are solutions to the equations for the lengths
+# arising from the cylinder diagram
+function realizable_lengths(rays, n)::Vector{Vector{Int}}
+  l = length(rays)
+  combinations = partition_degree(l, n)
+
+  result = Vector{Vector{Int}}()
+  exceeded = false
+  for combination in combinations
+    linear_combination = combination[1] * rays[1]
+    for i in 2:l
+      linear_combination = linear_combination + combination[i] * rays[i]
+      if sum(linear_combination) > n
+        exceeded = true
+        break
+      end
+    end
+    if exceeded == false
+      push!(result, linear_combination)
+    end
+    exceeded = false
+  end
+  return result
+end
+
+function realizable_lengths_of_cylinder_diagram(cyl::CylinderDiagram, n::Int)
+  equations = system_of_equations(cyl)
+  r = compute_rays(equations, cyl.separatrix_count)
+  return realizable_lengths(r, n)
+end
+
+# based on Donald E. Knuth The Art of Computer Programming Algorithm H p.300
+# implementation from Sage
+function product_gray_code(m::Vector{Int})
+  n = 0
+  k = 0
+  # assumes each radix is >= 2
+  new_m = Vector{Int}()
+  mm = Vector{Int}()
+  for i in m
+    if i <= 0
+      throw(ArgumentError("accepts only positive radices"))
+    elseif i >= 2
+      push!(new_m, i - 1)
+      push!(mm, k)
+      n += 1
+    end
+    k += 1
+  end
+
+  m = new_m
+  f = collect(1:(n + 1))
+  o = ones(Int, n)
+  a = zeros(Int, n)
+  results = Vector{Tuple{Int,Int}}()
+
+  j = f[1]
+  while j != n + 1
+    f[1] = 1
+    oo = o[j]
+    a[j] += oo
+    if a[j] == 0 || a[j] == m[j]
+      f[j] = f[j + 1]
+      f[j + 1] = j + 1
+      o[j] = -oo
+    end
+
+    push!(results, (mm[j] + 1, oo))
+    j = f[1]
+  end
+
+  return results
+end
+
+# get origamis from cylinder diagram and realizable lengths
+# from surface_dynamics by Vincent, Delecroix et al.
+function origami_from_cylinder_coordinates(
+  cyl_diagram::CylinderDiagram, lengths::Vector{Int}, heights::Vector{Int}
+)
+  # the total width of each cylinder is the sum of the lengths of the separatrices
+  widths = [sum(lengths[i + 1] for i in bot) for bot in cyl_diagram.bot]
+  areas = [heights[i] * widths[i] for i in 1:(cyl_diagram.cycles_count)]
+
+  v = [0]
+  for a in areas
+    push!(v, v[end] + a)
+  end
+
+  # prepare for twist
+  sep_bottom_pos = zeros(Int, cyl_diagram.separatrix_count)
+  for (i, bot) in enumerate(cyl_diagram.bot)
+    w = 0
+    for j in bot
+      sep_bottom_pos[j + 1] = v[i] + w
+      w += lengths[j + 1]
+    end
+  end
+
+  # horizontal permutation
+  lx = collect(1:v[end])
+  for i in 1:(cyl_diagram.cycles_count)
+    for j in v[i]:widths[i]:(v[i + 1] - 1)
+      lx[j + widths[i]] = j
+    end
+  end
+  lx = perm([x + 1 for x in lx])
+
+  ly = Int[]
+  for i in 1:(cyl_diagram.cycles_count)
+    append!(ly, (v[i] + widths[i]):(v[i + 1] - 1))
+    append!(ly, zeros(Int64, widths[i]))
+  end
+
+  for (i, top_seps) in enumerate(cyl_diagram.top)
+    top = []
```

Comment:
This creates a `Vector{Any}`, better would be to create a `Vector{FOO}` where `FOO` is a concrete type. I suspect here this would be `Int` here:
```suggestion
    top = Int[]
```